internal/invocation/invocation_service.go

```diff
@@ -17,68 +17,143 @@
 package invocation
 
 import (
+	"context"
 	"fmt"
+	"sync/atomic"
+	"time"
 
+	"github.com/hazelcast/hazelcast-go-client/hzerrors"
+	"github.com/hazelcast/hazelcast-go-client/internal/cb"
 	ilogger "github.com/hazelcast/hazelcast-go-client/internal/logger"
 	"github.com/hazelcast/hazelcast-go-client/internal/proto"
 )
 
+const (
+	ready   = 0
+	stopped = 1
+)
+
 type Handler interface {
 	Invoke(invocation Invocation) (groupID int64, err error)
 }
 
 type Service struct {
-	requestCh       <-chan Invocation
-	urgentRequestCh <-chan Invocation
-	responseCh      <-chan *proto.ClientMessage
+	requestCh       chan Invocation
+	urgentRequestCh chan Invocation
+	responseCh      chan *proto.ClientMessage
 	// removeCh carries correlationIDs to be removed
-	removeCh    <-chan int64
+	removeCh    chan int64
 	doneCh      chan struct{}
 	invocations map[int64]Invocation
 	handler     Handler
 	logger      ilogger.Logger
+	state       int32
 }
 
 func NewService(
-	requestCh <-chan Invocation,
-	urgentRequestCh <-chan Invocation,
-	responseCh <-chan *proto.ClientMessage,
-	removeCh <-chan int64,
 	handler Handler,
 	logger ilogger.Logger) *Service {
 	service := &Service{
-		requestCh:       requestCh,
-		urgentRequestCh: urgentRequestCh,
-		responseCh:      responseCh,
-		removeCh:        removeCh,
+		requestCh:       make(chan Invocation),
+		urgentRequestCh: make(chan Invocation),
+		responseCh:      make(chan *proto.ClientMessage),
+		removeCh:        make(chan int64),
 		doneCh:          make(chan struct{}),
 		invocations:     map[int64]Invocation{},
 		handler:         handler,
 		logger:          logger,
+		state:           ready,
 	}
 	go service.processIncoming()
 	return service
 }
 
 func (s *Service) Stop() {
+	if !atomic.CompareAndSwapInt32(&s.state, ready, stopped) {
+		return
+	}
 	close(s.doneCh)
 }
 
 func (s *Service) SetHandler(handler Handler) {
 	s.handler = handler
 }
 
+func (s *Service) SendRequest(ctx context.Context, inv Invocation) error {
+	if atomic.LoadInt32(&s.state) != ready {
```

Comment:
It looks like we don't need this state to be checked when writing to channels anymore. 
We will return from `<-s.doneCh` always when it is closed. 

It looks like it is only needed on `Stop` to make sure it is closed only once. And maybe to make it the purpose obvious ,we could use `once.Do`

Reply:
If you don't have a strong opinion about this, I'd say let's keep the state. Even if we use `sync.Once` we'd still need a field for that. Also other components work with state, so we stay more consistent.

Reply:
Lets just remove the checks from these methods then. As far as I see, they are not needed. Less code is always better. 
We just need check it on Stop